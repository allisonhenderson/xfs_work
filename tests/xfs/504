#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2018, Oracle and/or its affiliates.  All Rights Reserved.
#
# FS QA Test No.504
#
# simple parent pointer test
#

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
seqfull="$seqres.full"
status=1	# failure is the default!
trap "cleanup; exit \$status" 0 1 2 3 15

cleanup()
{
	cd /
	rm -f $tmp.*
}

filter_dbval()
{
	awk '{ print $4 }'
}

full()
{
    echo ""            >>$seqfull
    echo "*** $* ***"  >>$seqfull
    echo ""            >>$seqfull
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter
. ./common/reflink
. ./common/inject
. ./common/parent

# remove previous $seqres.full before test
rm -f $seqres.full

# Modify as appropriate
_supported_os Linux
_supported_fs xfs
_require_scratch
_require_xfs_io_error_injection "delayed_attr"

# real QA test starts here

# Create a directory tree using a protofile and
# make sure all inodes created have parent pointers

protofile=$tmp.proto

cat >$protofile <<EOF
DUMMY1
0 0
: root directory
d--777 3 1
: a directory
testfolder1 d--755 3 1
file1 ---755 3 1 /dev/null
$
: back in the root
testfolder2 d--755 3 1
file2 ---755 3 1 /dev/null
: done
$
EOF

if [ $? -ne 0 ]
then
    _fail "failed to create test protofile"
fi

echo "*** create FS"

rm -f $seqfull
_scratch_unmount >/dev/null 2>&1

_scratch_mkfs_xfs -f -n parent=1 -p $protofile >>$seqfull 2>&1 \
	|| _fail "mkfs failed"
echo "*** check FS"
_check_scratch_fs

echo "*** mount FS"
full " mount"
_scratch_mount >>$seqfull 2>&1 \
	|| _fail "mount failed"

testfolder1="testfolder1"
testfolder2="testfolder2"
file1="file1"
file2="file2"
file3="file3"
file4="file4"
file5="file5"
file1_ln="file1_link"

echo ""
echo "Create parent pointer test"
_verify_parent "$testfolder1" "$file1" "$testfolder1/$file1"

echo ""
echo "Move parent pointer test"
mv $SCRATCH_MNT/$testfolder1/$file1 $SCRATCH_MNT/$testfolder2/$file1
_verify_parent "$testfolder2" "$file1" "$testfolder2/$file1"

echo ""
echo "Hard link parent pointer test"
ln $SCRATCH_MNT/$testfolder2/$file1 $SCRATCH_MNT/$testfolder1/$file1_ln
_verify_parent "$testfolder1" "$file1_ln"  "$testfolder1/$file1_ln"
_verify_parent "$testfolder1" "$file1_ln"  "$testfolder2/$file1"
_verify_parent "$testfolder2" "$file1"     "$testfolder1/$file1_ln"
_verify_parent "$testfolder2" "$file1"     "$testfolder2/$file1"

echo ""
echo "Remove hard link parent pointer test"
ino="$(stat -c '%i' $SCRATCH_MNT/$testfolder2/$file1)"
rm $SCRATCH_MNT/$testfolder2/$file1
_verify_parent    "$testfolder1" "$file1_ln" "$testfolder1/$file1_ln"
_verify_no_parent "$file1" "$ino" "$testfolder1/$file1_ln"

echo ""
echo "Rename parent pointer test"
ino="$(stat -c '%i' $SCRATCH_MNT/$testfolder1/$file1_ln)"
mv $SCRATCH_MNT/$testfolder1/$file1_ln $SCRATCH_MNT/$testfolder1/$file2
_verify_parent    "$testfolder1" "$file2"    "$testfolder1/$file2"
_verify_no_parent "$file1_ln" "$ino" "$testfolder1/$file2"

echo ""
echo "Over write parent pointer test"
touch $SCRATCH_MNT/$testfolder2/$file3
_verify_parent    "$testfolder2" "$file3"    "$testfolder2/$file3"
ino="$(stat -c '%i' $SCRATCH_MNT/$testfolder2/$file3)"
mv -f $SCRATCH_MNT/$testfolder2/$file3 $SCRATCH_MNT/$testfolder1/$file2
_verify_parent    "$testfolder1" "$file2"    "$testfolder1/$file2"

echo "*** unmount FS"
full "umount"
_scratch_unmount >>$seqfull 2>&1 \
        || _fail "umount failed"

echo "*** done"
rm $seqfull
status=0

# success, all done
status=0
exit
