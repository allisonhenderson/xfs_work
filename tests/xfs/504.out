QA output created by 504
*** create FS
*** check FS
*** mount FS

Create parent pointer test
*** testfolder1 OK
*** testfolder1/file1 OK
*** testfolder1/file1 OK
*** Verified parent pointer: inode:16777344, name:file1, namelen:5
*** Parent pointer OK for child testfolder1/file1 inode:16777345

Move parent pointer test
*** testfolder2 OK
*** testfolder2/file1 OK
*** testfolder2/file1 OK
*** Verified parent pointer: inode:33613312, name:file1, namelen:5
*** Parent pointer OK for child testfolder2/file1 inode:16777345

Hard link parent pointer test
*** testfolder1 OK
*** testfolder1/file1_link OK
*** testfolder1/file1_link OK
*** Verified parent pointer: inode:16777344, name:file1_link, namelen:10
*** Parent pointer OK for child testfolder1/file1_link inode:16777345
*** testfolder1 OK
*** testfolder2/file1 OK
*** testfolder1/file1_link OK
*** Verified parent pointer: inode:16777344, name:file1_link, namelen:10
*** Parent pointer OK for child testfolder2/file1 inode:16777345
*** testfolder2 OK
*** testfolder1/file1_link OK
*** testfolder2/file1 OK
*** Verified parent pointer: inode:33613312, name:file1, namelen:5
*** Parent pointer OK for child testfolder1/file1_link inode:16777345
*** testfolder2 OK
*** testfolder2/file1 OK
*** testfolder2/file1 OK
*** Verified parent pointer: inode:33613312, name:file1, namelen:5
*** Parent pointer OK for child testfolder2/file1 inode:16777345

Remove hard link parent pointer test
*** testfolder1 OK
*** testfolder1/file1_link OK
*** testfolder1/file1_link OK
*** Verified parent pointer: inode:16777344, name:file1_link, namelen:10
*** Parent pointer OK for child testfolder1/file1_link inode:16777345
*** testfolder1/file1_link OK

Rename parent pointer test
*** testfolder1 OK
*** testfolder1/file2 OK
*** testfolder1/file2 OK
*** Verified parent pointer: inode:16777344, name:file2, namelen:5
*** Parent pointer OK for child testfolder1/file2 inode:16777345
*** testfolder1/file2 OK

Over write parent pointer test
*** testfolder2 OK
*** testfolder2/file3 OK
*** testfolder2/file3 OK
*** Verified parent pointer: inode:33613312, name:file3, namelen:5
*** Parent pointer OK for child testfolder2/file3 inode:33613314
*** testfolder1 OK
*** testfolder1/file2 OK
*** testfolder1/file2 OK
*** Verified parent pointer: inode:16777344, name:file2, namelen:5
*** Parent pointer OK for child testfolder1/file2 inode:33613314
*** unmount FS
*** done
